# Palpite
Cartola FC tips.

## Goal
To develop a model to predict how many points will an athlete score on Cartola FC.

The model is planned to be deployd on a web app and shall only use information available on the official Cartola FC API and on any popular betting API.

## Data Sources
### Cartola FC
The historical Cartola FC dataset used on this model is available on [github.com/henriquepgomide/caRtola](https://github.com/henriquepgomide/caRtola/tree/master/data).

### Betting Lines
The historical betting lines for Brasileirão Série A is available on [football-data.co.uk/brazil.php](https://www.football-data.co.uk/brazil.php).

Many thanks to the maintainers from both projects on making the data available for free to the world.

## Time Interval
The official Cartola FC API changed its format on 2018. Datasets before this year do not attend the expected formats, and therefore, won't be used.

The seasons used to build the model will be:

* 38 2018 rounds (complete season)
* 38 2019 rounds (complete season)
* 18 2020 rounds (almost half season)

It means:

* 2.5 seaons
* 94 rounds
* 940 matches

It seems to be enough and there is no need of trying to format previous years' datasets.

## Imports

In [1]:
import datetime
import itertools
import os
import pytz

import numpy as np
import pandas as pd

## Cartola FC
### Load data
Load data from 2018, 2019 and 2020 Cartola FC seasons.

In [2]:
def load_all_rounds(season):
    """ 
    Load all rounds data from a season to a pandas DataFrame.
    
    Also rename columns to a shorter name and introduce a year feature.
    """
    # Define file paths.
    file_path_list = [f"data//cartola//{season}//rodada-{i + 1}.csv" for i in range(38)]
    # Load datasets into a list.
    rounds = [
        pd.read_csv(file_path, index_col=0) for file_path in file_path_list if os.path.exists(file_path)
        ]
    # Concatenate the list.
    data = pd.concat(rounds, ignore_index=True)

    # Remove and repetive term from columns names.
    data.columns = [col.replace("atletas.", "") for col in data.columns]

    # Create year column.
    data["year"] = [season for _ in range(data.shape[0])]

    return data


def load_seasons(seasons):
    """ Load specified seasons into a same pandas DataFrame. """
    return pd.concat([load_all_rounds(year) for year in seasons], ignore_index=True)
    

cartola_data = load_seasons([2018, 2019, 2020])
cartola_data.head()

nome            slug         apelido  \
0         Matheus Ferraz Pereira  matheus-ferraz  Matheus Ferraz   
1          Willian Lanes de Lima            lima            Lima   
2           Rómulo Otero Vásquez           otero           Otero   
3           Diego Ribas da Cunha           diego           Diego   
4  Rodrigo Eduardo Costa Marinho     rodriguinho     Rodriguinho   

                                                foto  atleta_id  rodada_id  \
0  https://s.glbimg.com/es/sde/f/2018/03/17/6d461...      38632          1   
1  https://s.glbimg.com/es/sde/f/2018/03/17/3d9ef...      38506          1   
2  https://s.glbimg.com/es/sde/f/2017/04/03/9fe40...      83004          1   
3  https://s.glbimg.com/es/sde/f/2017/08/16/3ba37...      38909          1   
4  https://s.glbimg.com/es/sde/f/2018/03/20/c125f...      61033          1   

  clube_id posicao_id status_id  pontos_num  ...  FT  GS  CV  GC  PP  DP  \
0      AME        zag      Nulo         0.0  ... NaN NaN NaN NaN NaN NaN   
1      AME        zag      Nulo         0.0  ... NaN NaN NaN NaN NaN NaN   
2      ATL        mei  Provável        16.5  ... NaN NaN NaN NaN NaN NaN   
3      FLA        mei  Provável         0.8  ... NaN NaN NaN NaN NaN NaN   
4      COR        mei  Provável        16.5  ... NaN NaN NaN NaN NaN NaN   

   year  jogos_num  PI  DS  
0  2018        NaN NaN NaN  
1  2018        NaN NaN NaN  
2  2018        NaN NaN NaN  
3  2018        NaN NaN NaN  
4  2018        NaN NaN NaN  

[5 rows x 36 columns]

### Keep relevant data
This model do not aim to use athletes scouts (stats). They are all removed. 

Redundant features, like athlete id, name, nickname, slug are also removed.

Price-related features are removed since I want the model to completely disregard prices (unbiased to pricing).

In [3]:
to_keep = [
    'year',
    'apelido',  # Athlete nickname.
    'rodada_id',  # Round number.
    'clube.id.full.name',  # Club full name.
    'posicao_id',  # Position ID.
    'status_id',  # Status ID.
    'media_num',  # Mean number.
    'pontos_num',  # Points number.
]

cartola_data = cartola_data[to_keep]
cartola_data.tail()

year   apelido  rodada_id clube.id.full.name posicao_id status_id  \
74588  2020     Rildo         18                284        mei      Nulo   
74589  2020    Danilo         18                373        zag      Nulo   
74590  2020  Alvarado         18                293        mei      Nulo   
74591  2020   Romildo         18                263        mei      Nulo   
74592  2020   Nicolas         18                373        lat  Provável   

       media_num  pontos_num  
74588       0.15         0.0  
74589       0.00         0.0  
74590       0.82         0.0  
74591       0.00         0.0  
74592       3.07         0.5

### Translate
Code should preferable be written in English. 


Even though it is a brazilian fantasy game, with a brazilian target audience, 
you never know when someone else may be interested in reading your code for whatever the purpose it may be. 
So it is a nice idea to always keep it in English.

In [4]:
# Rename columns.
cartola_data = cartola_data.rename({
    "apelido": "name",
    "rodada_id": "round",
    "clube.id.full.name": "club",
    "posicao_id": "position",
    "status_id": "status",
    "pontos_num": "points",
    "media_num": "mean",
}, axis=1)

# Rename values from the column position.
cartola_data["position"] = cartola_data["position"].replace({
    "zag": "defender",
    "mei": "midfielder",
    "ata": "forward",
    "gol": "goalkeeper",
    "lat": "fullback",
    "tec": "coach",
})

# Rename values from the column status.
cartola_data["status"] = cartola_data["status"].replace({
    "Nulo": "null",
    "Provável": "expected",
    "Contudido": "injured",
    "Dúvida": "doubt",
    "Suspenso": "suspended",
})

cartola_data.tail()

year      name  round club    position    status  mean  points
74588  2020     Rildo     18  284  midfielder      null  0.15     0.0
74589  2020    Danilo     18  373    defender      null  0.00     0.0
74590  2020  Alvarado     18  293  midfielder      null  0.82     0.0
74591  2020   Romildo     18  263  midfielder      null  0.00     0.0
74592  2020   Nicolas     18  373    fullback  expected  3.07     0.5

### Clubs
It looks like there is pertubation in the `club` column.

Let's further invertigate it and decide how to fix it.

In [5]:
cartola_data["club"].unique()

array(['América-MG', 'Atlético-MG', 'Flamengo', 'Corinthians',
       'Atlético-PR', 'Botafogo', 'Bahia', 'São Paulo', 'Chapecoense',
       'Sport', 'Vasco', 'Fluminense', 'Internacional', 'Vitória',
       'Cruzeiro', 'Grêmio', 'Ceará', 'Paraná', 'Palmeiras', 'Santos',
       'Avaí', 'Athlético-PR', 'Fortaleza', 'CSA', 'Goiás', 'Atlético-GO',
       'Coritiba', 'Bragantino', '373', '293', '282', '284', '276', '265',
       '266', '356', '264', '354', '267', '263', '290', '292', '285',
       '275', '280'], dtype=object)

Some clubs names are represeted by its full name, while others are using the ID. To be coherent, I'll convert those IDs into names.

Since the names are case sensitive and have special characters that may produce errors later on, why not use IDs only? 
It is because the ID system changed throughout the years. It means that a same club may have different IDs for the year 2018 and 2019. It could produce erros that are harder to debug.

Even though full names are also prone to error, they are easier to debug, since they are written in natural language.

In order to correlate IDs to names, first it is important to understant in which year it is happening to use the most appropiate mapping.


In [6]:

def is_int(value):
    """ Check if a value is integer. """
    try:
        int(value)
        return True
    except ValueError:
        return False


def count_int(sequence):
    """ Count the amount of integers in a sequence. """
    return len([value for value in sequence if is_int(value)])


print(f"There are {count_int(cartola_data[cartola_data['year'] == 2018]['club'])} IDs in the year 2018")
print(f"There are {count_int(cartola_data[cartola_data['year'] == 2019]['club'])} IDs in the year 2019")
print(f"There are {count_int(cartola_data[cartola_data['year'] == 2020]['club'])} IDs in the year 2020")

There are 0 IDs in the year 2018
There are 0 IDs in the year 2019
There are 4396 IDs in the year 2020


All of the IDs are from the 2020 year. So I'll use the mapping from this year. The clubs dataset was retrieved from the official Cartola FC API.

In [7]:
clubs_data = pd.read_csv(r"data\cartola\clubes.csv", index_col=0)
clubs_data.head()

id      nome abreviacao  \
1        1    Outros        OUT   
1349  1349  Ipatinga        IPA   
1371  1371    Cuiabá        CUI   
1390  1390     Icasa        ICA   
2190  2190     Oeste        OES   

                                                escudos nome_fantasia  
1     {'60x60': 'https://s.glbimg.com/es/sde/f/organ...        Outros  
1349  {'60x60': 'https://s.glbimg.com/es/sde/f/organ...      Ipatinga  
1371  {'60x60': 'https://s.glbimg.com/es/sde/f/organ...        Cuiabá  
1390  {'60x60': 'https://s.glbimg.com/es/sde/f/organ...         Icasa  
2190  {'60x60': 'https://s.glbimg.com/es/sde/f/organ...         Oeste

#### Translate Clubs Dataset
For the same reason as before, the dataset needs to be translated.

In [8]:
# Rename columns.
clubs_data = clubs_data.rename({
    "nome": "name",
    "abreviacao": "abbreviation",
    "escudos": "logos",
    "nome_fantasia": "fanstasy_name",
}, axis=1)

clubs_data.head()

id      name abbreviation  \
1        1    Outros          OUT   
1349  1349  Ipatinga          IPA   
1371  1371    Cuiabá          CUI   
1390  1390     Icasa          ICA   
2190  2190     Oeste          OES   

                                                  logos fanstasy_name  
1     {'60x60': 'https://s.glbimg.com/es/sde/f/organ...        Outros  
1349  {'60x60': 'https://s.glbimg.com/es/sde/f/organ...      Ipatinga  
1371  {'60x60': 'https://s.glbimg.com/es/sde/f/organ...        Cuiabá  
1390  {'60x60': 'https://s.glbimg.com/es/sde/f/organ...         Icasa  
2190  {'60x60': 'https://s.glbimg.com/es/sde/f/organ...         Oeste

In [9]:
def apply_mapping(value, mapping):
    """ Apply a mapping to column values. """
    if value in mapping:
        return mapping[value]
    return value



id_to_name_mapping = {str(row.id): row.name for row in clubs_data.itertuples()}

cartola_data["club"] = cartola_data["club"].apply(apply_mapping, mapping=id_to_name_mapping)
cartola_data["club"].unique()

array(['América-MG', 'Atlético-MG', 'Flamengo', 'Corinthians',
       'Atlético-PR', 'Botafogo', 'Bahia', 'São Paulo', 'Chapecoense',
       'Sport', 'Vasco', 'Fluminense', 'Internacional', 'Vitória',
       'Cruzeiro', 'Grêmio', 'Ceará', 'Paraná', 'Palmeiras', 'Santos',
       'Avaí', 'Athlético-PR', 'Fortaleza', 'CSA', 'Goiás', 'Atlético-GO',
       'Coritiba', 'Bragantino'], dtype=object)

We are almost done with the clubs. The last thing that needs to be fixed is the name from Athlético-PR.

This team changes its name in recent years and because of it, is duplicated.

In [10]:
cartola_data["club"] = cartola_data["club"].replace("Atlético-PR", "Athlético-PR")
cartola_data["club"].unique()

array(['América-MG', 'Atlético-MG', 'Flamengo', 'Corinthians',
       'Athlético-PR', 'Botafogo', 'Bahia', 'São Paulo', 'Chapecoense',
       'Sport', 'Vasco', 'Fluminense', 'Internacional', 'Vitória',
       'Cruzeiro', 'Grêmio', 'Ceará', 'Paraná', 'Palmeiras', 'Santos',
       'Avaí', 'Fortaleza', 'CSA', 'Goiás', 'Atlético-GO', 'Coritiba',
       'Bragantino'], dtype=object)

### Data Transformation
There is still one big problem.

The following columns vary from game to game. 

* `status`
* `points`
* `mean`

Status is available before the match to the user, while mean is available after the match. In the next round the user will have status for the next match, and the mean from the last match.

Therefore, for a model to be useful for the user, it should consider next match status, and last match mean.

Naturally, for the first match, mean will be null. Also, Notice that points won't be transformed since it is the target feature.

In [11]:
# Duplicate data and reduce the round feature
antecipated_data = cartola_data.copy()
antecipated_data["round"] = [round_ - 1 for round_ in antecipated_data["round"]]

# Join on features that don't change
join_on = ['year', 'name', 'round', 'club', 'position']
cartola_data = pd.merge(
    left=antecipated_data, 
    right=cartola_data, 
    on=join_on, 
    how="left", 
    suffixes=("_current", "_previous")  # Translate to _current, _previous
    )

# Reset rounds.
cartola_data["round"] = [round + 1 for round in cartola_data["round"]]

# Remove columns that won't be no longer used.
cartola_data = cartola_data.drop(columns=["status_previous", "points_previous", "mean_current"])

# Remove suffixes
cartola_data.columns = [col.replace("_previous", "").replace("_current", "") for col in cartola_data.columns]

# Transform mean NaN in 0
cartola_data["mean"] = cartola_data["mean"].fillna(0)

cartola_data.head()

year            name  round         club    position    status  points  \
0  2018  Matheus Ferraz      1   América-MG    defender      null     0.0   
1  2018            Lima      1   América-MG    defender      null     0.0   
2  2018           Otero      1  Atlético-MG  midfielder  expected    16.5   
3  2018           Diego      1     Flamengo  midfielder  expected     0.8   
4  2018     Rodriguinho      1  Corinthians  midfielder  expected    16.5   

   mean  
0   0.0  
1   0.0  
2   0.0  
3   0.0  
4   0.0

### Dates
We should introduce a dates feature. This is important in order to merge with the betting lines dataset later on.

I'll concatenate the matches datasets from each year.

In [12]:
def load_date_dataset(year):
    """ Load dates dataset. """
    data = pd.read_csv(f"data\\cartola\\{year}\\{year}_partidas.csv")
    data["year"] = [year for _ in range(data.shape[0])]
    return data


# Load dates datasets.
years = [2018, 2019, 2020]
matches_data = pd.concat([load_date_dataset(year) for year in years], ignore_index=True)

print(f"There are {count_int(matches_data[matches_data['year'] == 2018]['home_team'])} IDs in the year 2018")
print(f"There are {count_int(matches_data[matches_data['year'] == 2019]['home_team'])} IDs in the year 2019")
print(f"There are {count_int(matches_data[matches_data['year'] == 2020]['home_team'])} IDs in the year 2020")

matches_data.head()

There are 0 IDs in the year 2018
There are 380 IDs in the year 2019
There are 180 IDs in the year 2020


game  round                date           home_team  score  \
0   1.0      1  14/04/2018 - 16:00       Cruzeiro - MG  0 x 1   
1   2.0      1  15/04/2018 - 19:00       Atlético - PR  5 x 1   
2   3.0      1  15/04/2018 - 11:00        América - MG  3 x 0   
3   4.0      1  14/04/2018 - 19:00        Vitória - BA  2 x 2   
4   5.0      1  15/04/2018 - 16:00  Vasco da Gama - RJ  2 x 1   

          away_team                                arena  year  home_score  \
0       Grêmio - RS       Mineirão - Belo Horizonte - MG  2018         NaN   
1  Chapecoense - SC     Arena da Baixada - Curitiba - PR  2018         NaN   
2        Sport - PE  Independência - Belo Horizonte - MG  2018         NaN   
3     Flamengo - RJ      Manoel Barradas - Salvador - BA  2018         NaN   
4     Atlético - MG   São Januário - Rio de Janeiro - RJ  2018         NaN   

   away_score  
0         NaN  
1         NaN  
2         NaN  
3         NaN  
4         NaN

This dataset also have some problems that should be fixed:

* Dates are in different format depending on the year.
* Team names are in various formats.

To fix teams names I'll use the `time_ids.csv` dataset. It as some other teams names mapping.


In [13]:
alternative_clubs_data = pd.read_csv(r"data\cartola\times_ids.csv")
alternative_clubs_data.head()

nome.cbf nome.cartola        nome.completo  cod.older  cod.2017  \
0   América - MG   América-MG           America MG        327       327   
1   America - RN  Atlético-RN           America RN        200       200   
2  Atlético - GO  Atlético-GO          Atletico GO        201       373   
3  Atlético - MG  Atlético-MG     Atletico Mineiro        282       282   
4  Atlético - PR  Atlético-PR  Atletico Paranaense        293       293   

   cod.2018   id abreviacao  \
0       327  327        AME   
1         1  200        OUT   
2       373  373        ATL   
3       282  282        ATL   
4       293  293        ATL   

                                       escudos.60x60  \
0  https://s.glbimg.com/es/sde/f/organizacoes/201...   
1                                                NaN   
2                                                NaN   
3  https://s.glbimg.com/es/sde/f/equipes/2017/11/...   
4  https://s.glbimg.com/es/sde/f/equipes/2015/06/...   

                                       escudos.45x45  \
0  https://s.glbimg.com/es/sde/f/organizacoes/201...   
1                                                NaN   
2                                                NaN   
3  https://s.glbimg.com/es/sde/f/equipes/2017/11/...   
4  https://s.glbimg.com/es/sde/f/equipes/2015/06/...   

                                       escudos.30x30  
0  https://s.glbimg.com/es/sde/f/organizacoes/201...  
1                                                NaN  
2                                                NaN  
3  https://s.glbimg.com/es/sde/f/equipes/2017/11/...  
4  https://s.glbimg.com/es/sde/f/equipes/2015/06/...

In [14]:
# Translate
alternative_clubs_data = alternative_clubs_data.rename({
    "nome.cbf": "cbf_name",  # CBF is the brazilian soccer confederation.
    "nome.cartola": "cartola_name",  # Cartola FC name.
    "nome.completo": "full_name",
    "cod_older": "id_old",  # ID as before 2017.
    "cod.2017": "id_2017",  # ID as 2017
    "cod.2018": "id_2018",  # ID as 2018
    "id": "id_2019",  # ID as 2019
    "abreviacao": "abbreviation",
    "escudo.60x60": "logo_60x60",
    "escudo.45x45": "logo_45x45",
    "escudo.30x30": "logo_30x30",
}, axis=1)

alternative_clubs_data["cartola_name"] = alternative_clubs_data["cartola_name"].replace(
    "Atlético-PR", "Athlético-PR"
    )

alternative_clubs_data.head()

cbf_name  cartola_name            full_name  cod.older  id_2017  \
0   América - MG    América-MG           America MG        327      327   
1   America - RN   Atlético-RN           America RN        200      200   
2  Atlético - GO   Atlético-GO          Atletico GO        201      373   
3  Atlético - MG   Atlético-MG     Atletico Mineiro        282      282   
4  Atlético - PR  Athlético-PR  Atletico Paranaense        293      293   

   id_2018  id_2019 abbreviation  \
0      327      327          AME   
1        1      200          OUT   
2      373      373          ATL   
3      282      282          ATL   
4      293      293          ATL   

                                       escudos.60x60  \
0  https://s.glbimg.com/es/sde/f/organizacoes/201...   
1                                                NaN   
2                                                NaN   
3  https://s.glbimg.com/es/sde/f/equipes/2017/11/...   
4  https://s.glbimg.com/es/sde/f/equipes/2015/06/...   

                                       escudos.45x45  \
0  https://s.glbimg.com/es/sde/f/organizacoes/201...   
1                                                NaN   
2                                                NaN   
3  https://s.glbimg.com/es/sde/f/equipes/2017/11/...   
4  https://s.glbimg.com/es/sde/f/equipes/2015/06/...   

                                       escudos.30x30  
0  https://s.glbimg.com/es/sde/f/organizacoes/201...  
1                                                NaN  
2                                                NaN  
3  https://s.glbimg.com/es/sde/f/equipes/2017/11/...  
4  https://s.glbimg.com/es/sde/f/equipes/2015/06/...

In [15]:
# Create a dictionary from the dataset.
cbf_to_cartola_mapping = {
    row.cbf_name: row.cartola_name for row in alternative_clubs_data.itertuples()
    }
id_to_name_mapping_2019 = {
    row.id_2019: row.cartola_name for row in alternative_clubs_data.itertuples()
    }
id_to_name_mapping_2020 = {int(row.id): row.name for row in clubs_data.itertuples()}

# Apply CBF name mapping function.
year_to_mapping = (
    ([2018, 2019, 2020], cbf_to_cartola_mapping,),
    ([2019], id_to_name_mapping_2019,),
    ([2020], id_to_name_mapping_2020,),
)

for years, mapping in year_to_mapping:
    # Filter.
    matches_data_year = matches_data[matches_data["year"].isin(years)]
    
    # Map.
    matches_data_year["home_team"] = matches_data_year["home_team"].apply(
        apply_mapping, mapping=mapping
        )
    matches_data_year["away_team"] = matches_data_year["away_team"].apply(
        apply_mapping, mapping=mapping
        )

    # Apply.
    matches_data[matches_data["year"].isin(years)] = matches_data_year

matches_data["home_team"].unique()

array(['Cruzeiro', 'Athlético-PR', 'América-MG', 'Vitória', 'Vasco',
       'Botafogo', 'São Paulo', 'Santos', 'Corinthians', 'Internacional',
       'Atlético-MG', 'Paraná', 'Chapecoense', 'Bahia', 'Fluminense',
       'Flamengo', 'Ceará-SC', 'Sport', 'Palmeiras', 'Grêmio', 'Avaí',
       'CSA', 'Fortaleza', 'Goiás', 'Coritiba', 'Bragantino', 'Ceará',
       'Atlético-GO'], dtype=object)

Ceará appears twice as *Ceará-SC* and *Ceará*. It should be fixed.

In [16]:
matches_data["home_team"] = matches_data["home_team"].replace("Ceará-SC", "Ceará")
matches_data["away_team"] = matches_data["away_team"].replace("Ceará-SC", "Ceará")

matches_data["home_team"].unique()

array(['Cruzeiro', 'Athlético-PR', 'América-MG', 'Vitória', 'Vasco',
       'Botafogo', 'São Paulo', 'Santos', 'Corinthians', 'Internacional',
       'Atlético-MG', 'Paraná', 'Chapecoense', 'Bahia', 'Fluminense',
       'Flamengo', 'Ceará', 'Sport', 'Palmeiras', 'Grêmio', 'Avaí', 'CSA',
       'Fortaleza', 'Goiás', 'Coritiba', 'Bragantino', 'Atlético-GO'],
      dtype=object)

We're done with teams, now is time to clean up dates.

In [17]:
def interpret_date(string, patterns, timezone=None):
    """ Interpret some specific patterns of date. """
    for pattern in patterns:
        try:
            dt = datetime.datetime.strptime(string, pattern)
            if timezone:
                dt = dt + datetime.timedelta(hours=timezone)
            return pd.to_datetime(dt.date())
        except ValueError:
            pass
    raise ValueError("Couldn't match any date pattern.")


# Interpret string ad datetime.
matches_data["date"] = matches_data["date"].apply(
    interpret_date, patterns=["%d/%m/%Y - %H:%M", "%Y-%m-%d"])

matches_data.head(5)

game  round       date     home_team  score    away_team  \
0   1.0      1 2018-04-14      Cruzeiro  0 x 1       Grêmio   
1   2.0      1 2018-04-15  Athlético-PR  5 x 1  Chapecoense   
2   3.0      1 2018-04-15    América-MG  3 x 0        Sport   
3   4.0      1 2018-04-14       Vitória  2 x 2     Flamengo   
4   5.0      1 2018-04-15         Vasco  2 x 1  Atlético-MG   

                                 arena  year  home_score  away_score  
0       Mineirão - Belo Horizonte - MG  2018         NaN         NaN  
1     Arena da Baixada - Curitiba - PR  2018         NaN         NaN  
2  Independência - Belo Horizonte - MG  2018         NaN         NaN  
3      Manoel Barradas - Salvador - BA  2018         NaN         NaN  
4   São Januário - Rio de Janeiro - RJ  2018         NaN         NaN

### Merge Cartola FC and matches datasets
Now that everything looks fine with the matches dataset, it is ready to be merged on the Cartola FC dataset.

In [18]:
# Rename the matches dataset to attend the cartola dataset features names. This is required for merging.
matches_renamed = matches_data.rename({
    "home_team": "club_home",
    "away_team": "club_away"
    }, axis=1)

# We need to do it in two steps. First merge searching for the club in home teams. 
# Then do it again for away teams.

for club_type in ["home", "away"]:
    matches_renamed["club"] = matches_renamed[f"club_{club_type}"]

    # Filter to merge only desired features.
    matches_renamed_filtered = matches_renamed[["round", "year", "club", "date"]]

    merge_on = ["round", "year", "club"]
    cartola_data = pd.merge(cartola_data, matches_renamed_filtered, on=merge_on, how="left")

# Unify both new columns in a single column.
cartola_data["date"] = np.where(
    pd.isna(cartola_data["date_x"]),
    cartola_data["date_y"],
    cartola_data["date_x"]
    )

# Remove date_x and date_y.
cartola_data = cartola_data.drop(columns=["date_x", "date_y"])

cartola_data.tail()

year      name  round          club    position    status  points  \
74757  2020     Rildo     18        Grêmio  midfielder      null     0.0   
74758  2020    Danilo     18   Atlético-GO    defender      null     0.0   
74759  2020  Alvarado     18  Athlético-PR  midfielder      null     0.0   
74760  2020   Romildo     18      Botafogo  midfielder      null     0.0   
74761  2020   Nicolas     18   Atlético-GO    fullback  expected     0.5   

       mean       date  
74757  0.15 2020-10-25  
74758  0.00 2020-10-25  
74759  0.82 2020-10-25  
74760  0.00 2020-10-25  
74761  3.24 2020-10-25

Check if all columns received a date value.

In [19]:
# Count NaNs in each row.
cartola_data.shape[0] - cartola_data.count()

year          0
name          0
round         0
club          0
position      0
status        0
points        0
mean          0
date        458
dtype: int64

There are still 458 rows without date. They needed to be investigated.

In [20]:
cartola_data[pd.isna(cartola_data["date"])].groupby(["year", "round", "club"]).first()

name    position     status  points  mean  \
year round club                                                                 
2020 1     Atlético-GO          Jorginho  midfielder   expected     0.0  0.00   
           Bahia                 Juninho    defender   expected     0.0  0.00   
           Botafogo                Honda  midfielder   expected     0.0  0.00   
           Corinthians              Luan  midfielder   expected     0.0  0.00   
     11    Athlético-PR           Santos  goalkeeper   expected     0.0  1.30   
           Bahia                Gilberto     forward   expected     0.0  2.61   
           Flamengo      Everton Ribeiro  midfielder   expected     0.0  3.33   
           Goiás            Rafael Moura     forward   expected     0.0  0.00   
           São Paulo      Fernando Diniz       coach   expected     0.0  0.00   
     12    Sport                 Hernane     forward   expected     0.0  0.85   
     18    Corinthians                Jô     forward       null     0.0  2.51   
           Vasco          Leandro Castan    defender  suspended     0.0  3.22   

                        date  
year round club               
2020 1     Atlético-GO   NaT  
           Bahia         NaT  
           Botafogo      NaT  
           Corinthians   NaT  
     11    Athlético-PR  NaT  
           Bahia         NaT  
           Flamengo      NaT  
           Goiás         NaT  
           São Paulo     NaT  
     12    Sport         NaT  
     18    Corinthians   NaT  
           Vasco         NaT

It looks likes those are games that really didn't happen. 2020 was a year with a lot of delayed games due to the COVID-19 pandemic.

In [21]:
cartola_data = cartola_data.dropna()

# Count NaNs in each row.
cartola_data.shape[0] - cartola_data.count()

year        0
name        0
round       0
club        0
position    0
status      0
points      0
mean        0
date        0
dtype: int64

## Betting Lines
### Load data
Load betting lines from 2018, 2019 and 2020 seasons.

In [22]:
bet_data = pd.read_csv(r"data\betting\historical_betting_lines.csv")
bet_data.head()

Country   League  Season        Date   Time          Home         Away   HG  \
0  Brazil  Serie A    2012  19/05/2012  22:30     Palmeiras   Portuguesa  1.0   
1  Brazil  Serie A    2012  19/05/2012  22:30  Sport Recife  Flamengo RJ  1.0   
2  Brazil  Serie A    2012  20/05/2012  01:00   Figueirense      Nautico  2.0   
3  Brazil  Serie A    2012  20/05/2012  20:00   Botafogo RJ    Sao Paulo  4.0   
4  Brazil  Serie A    2012  20/05/2012  20:00   Corinthians   Fluminense  0.0   

    AG Res    PH    PD    PA  MaxH  MaxD  MaxA  AvgH  AvgD  AvgA  
0  1.0   D  1.75  3.86  5.25  1.76  3.87  5.31  1.69  3.50  4.90  
1  1.0   D  2.83  3.39  2.68  2.83  3.42  2.70  2.59  3.23  2.58  
2  1.0   H  1.60  4.04  6.72  1.67  4.05  7.22  1.59  3.67  5.64  
3  2.0   H  2.49  3.35  3.15  2.49  3.39  3.15  2.35  3.26  2.84  
4  1.0   A  1.96  3.53  4.41  1.96  3.53  4.41  1.89  3.33  3.89

In [23]:
# Filter seasons.
bet_data = bet_data[bet_data["Season"].isin([2018, 2019, 2020])]
bet_data.head()

Country   League  Season        Date   Time         Home          Away  \
2280  Brazil  Serie A    2018  14/04/2018  20:00     Cruzeiro        Gremio   
2281  Brazil  Serie A    2018  14/04/2018  23:00      Vitoria   Flamengo RJ   
2282  Brazil  Serie A    2018  15/04/2018  01:00       Santos         Ceara   
2283  Brazil  Serie A    2018  15/04/2018  15:00   America MG  Sport Recife   
2284  Brazil  Serie A    2018  15/04/2018  20:00  Corinthians    Fluminense   

       HG   AG Res    PH    PD    PA  MaxH  MaxD  MaxA  AvgH  AvgD  AvgA  
2280  0.0  1.0   A  1.96  3.29  4.87  2.11  3.30  4.87  1.98  3.16  4.15  
2281  2.0  2.0   D  3.56  3.41  2.22  3.75  3.42  2.24  3.40  3.31  2.14  
2282  2.0  0.0   H  1.57  4.13  6.76  1.60  4.14  7.10  1.55  3.88  6.12  
2283  3.0  0.0   H  1.98  3.35  4.51  2.05  3.42  4.51  1.96  3.27  4.01  
2284  2.0  1.0   H  1.67  3.68  6.55  1.68  3.74  7.00  1.62  3.55  6.03

### Keep relevant rows
We are only interested in information that is available before the game.

We take as premise that the betting odds are accurate, and we shall not try to predict whenever they will fail. This would make the model over-complicated.

With that in mind, we only need the odds and the minimum information required to merge them with the Cartola FC dataset.

#### About the odds
We will use average odds instead of maximum or from a single source to avoid outliers or bad decisions by a specific booker.

In [24]:
to_keep = ["Season", "Date", "Time", "Home", "Away", "AvgH", "AvgD", "AvgA"]
bet_data = bet_data[to_keep]
bet_data.head()

Season        Date   Time         Home          Away  AvgH  AvgD  AvgA
2280    2018  14/04/2018  20:00     Cruzeiro        Gremio  1.98  3.16  4.15
2281    2018  14/04/2018  23:00      Vitoria   Flamengo RJ  3.40  3.31  2.14
2282    2018  15/04/2018  01:00       Santos         Ceara  1.55  3.88  6.12
2283    2018  15/04/2018  15:00   America MG  Sport Recife  1.96  3.27  4.01
2284    2018  15/04/2018  20:00  Corinthians    Fluminense  1.62  3.55  6.03

### Fit data format into Cartola FC dataset standards

Everything is in a different format related to the Cartola FC dataset. For a succesful merge this dataset needs to be reformatted.

In [25]:
# Concatenate Date and Time
bet_data["Date"] = [f"{row.Date} {row.Time}" for row in bet_data.itertuples()]
bet_data = bet_data.drop(columns=["Time"])

# Rename columns.
bet_data = bet_data.rename({
    "Season": "year",
    "Date": "date",
    "Home": "home",
    "Away": "away",
    "AvgH": "home_odds",
    "AvgD": "draw_odds",
    "AvgA": "away_odds",
}, axis=1)

# Interpret string ad datetime.
bet_data["date"] = bet_data["date"].apply(
    interpret_date, patterns=["%d/%m/%Y %H:%M"], timezone=-4,
    )

bet_data.head()

year       date         home          away  home_odds  draw_odds  \
2280  2018 2018-04-14     Cruzeiro        Gremio       1.98       3.16   
2281  2018 2018-04-14      Vitoria   Flamengo RJ       3.40       3.31   
2282  2018 2018-04-14       Santos         Ceara       1.55       3.88   
2283  2018 2018-04-15   America MG  Sport Recife       1.96       3.27   
2284  2018 2018-04-15  Corinthians    Fluminense       1.62       3.55   

      away_odds  
2280       4.15  
2281       2.14  
2282       6.12  
2283       4.01  
2284       6.03

And now for something completely different... fix clubs names.

This time is more challenging because it doesn't seem to match any pattern used before. I will have to carefully create a new mapping.

In [26]:
bet_data["home"].unique()

array(['Cruzeiro', 'Vitoria', 'Santos', 'America MG', 'Corinthians',
       'Internacional', 'Vasco', 'Atletico-PR', 'Botafogo RJ',
       'Sao Paulo', 'Bahia', 'Flamengo RJ', 'Parana', 'Atletico-MG',
       'Ceara', 'Chapecoense-SC', 'Fluminense', 'Palmeiras', 'Gremio',
       'Sport Recife', 'Athletico-PR', 'CSA', 'Avai', 'Fortaleza',
       'Goias', 'Coritiba', 'Bragantino', 'Atletico GO'], dtype=object)

To be corrected:

* No accentuation. 
* Some teams like Botafogo and Flamengo have the states or city appended to end.
* Some teams uses dashs and other spaces.

In [27]:
teams_names_mapping = {
    "Vitoria": "Vitória",
    "America MG": "América-MG",
    "Sao Paulo": "São Paulo",
    "Flamengo RJ": "Flamengo",
    "Botafogo RJ": "Botafogo",
    "Parana": "Paraná",
    "Atletico-MG": "Atlético-MG",
    "Ceara": "Ceará",
    "Chapecoense-SC": "Chapecoense",
    "Gremio": "Grêmio",
    "Sport Recife": "Sport",
    "Athletico-PR": "Athlético-PR",
    "Atletico-PR": "Athlético-PR",
    "Avai": "Avaí",
    "Goias": "Goiás",
    "Atletico GO": "Atlético-GO",

}

# Apply mapping.
for col in ["home", "away"]:
    bet_data[col] = bet_data[col].apply(apply_mapping, mapping=teams_names_mapping)

    # Check that new columns names are the same from Cartola FC dataset.
    if all([team in cartola_data["club"].unique() for team in bet_data[col].unique()]):
        print(f"Formatting was succesfull on column {col}.")
    else:
        print(f"Something wrong with column {col} formatting.")

bet_data.head()

Formatting was succesfull on column home.
Formatting was succesfull on column away.


year       date         home        away  home_odds  draw_odds  \
2280  2018 2018-04-14     Cruzeiro      Grêmio       1.98       3.16   
2281  2018 2018-04-14      Vitória    Flamengo       3.40       3.31   
2282  2018 2018-04-14       Santos       Ceará       1.55       3.88   
2283  2018 2018-04-15   América-MG       Sport       1.96       3.27   
2284  2018 2018-04-15  Corinthians  Fluminense       1.62       3.55   

      away_odds  
2280       4.15  
2281       2.14  
2282       6.12  
2283       4.01  
2284       6.03

## Merge Both Datasets
Finally the time has come to merge both datasets. 

In [28]:
# There may be problems with timezone, so consider an one day tolerance.
tol = pd.Timedelta(value=1, unit="day")
# merge_asof requires sorting the left variable.
cartola_data = cartola_data.sort_values("date", ignore_index=True)

# First merge home games.
bet_data["club"] = bet_data["home"]
home_games = pd.merge_asof(cartola_data, bet_data, on="date", by="club", direction="nearest", tolerance=tol)#, how="left")

# First merge away games.
bet_data["club"] = bet_data["away"]
away_games = pd.merge_asof(cartola_data, bet_data, on="date", by="club", direction="nearest", tolerance=tol)#, how="left")

cartola_data["win"] = np.where(
    pd.isna(home_games["home_odds"]),
    away_games["away_odds"],
    home_games["home_odds"]
    )

cartola_data["lose"] = np.where(
    pd.isna(home_games["away_odds"]),
    away_games["home_odds"],
    home_games["away_odds"]
)

cartola_data["draw"] = np.where(
    pd.isna(home_games["draw_odds"]),
    away_games["draw_odds"],
    home_games["draw_odds"]
)

cartola_data.head()

year        name  round      club    position      status  points  mean  \
0  2018   Pedro Ken      1     Ceará  midfielder  Contundido     0.0   0.0   
1  2018  Léo Xavier      1   Vitória    defender        null     0.0   0.0   
2  2018      Thuler      1  Flamengo    defender        null     0.0   0.0   
3  2018      Ramiro      1    Grêmio  midfielder    expected     2.9   0.0   
4  2018     Alisson      1    Grêmio  midfielder        null     0.0   0.0   

        date   win  lose  draw  
0 2018-04-14  6.12  1.55  3.88  
1 2018-04-14  3.40  2.14  3.31  
2 2018-04-14  2.14  3.40  3.31  
3 2018-04-14  4.15  1.98  3.16  
4 2018-04-14  4.15  1.98  3.16

### Check NaNs

Let's check if the merger was success.

In [29]:
# NaNs ratio in each row.
(cartola_data.shape[0] - cartola_data.count()) / cartola_data.shape[0]

year        0.000000
name        0.000000
round       0.000000
club        0.000000
position    0.000000
status      0.000000
points      0.000000
mean        0.000000
date        0.000000
win         0.011143
lose        0.011143
draw        0.011143
dtype: float64

The merge was unsuccesful for around 1% of occurences. This is negligible.

### Check Merging
This merger is the most important action so far. It is crutial that it was done correctly, otherwise the model will be completely wrong. It is a good idea to check it.

In [30]:
# Extract a sample from Cartola FC dataframe.
sample = cartola_data.sample(1).iloc[0]

# Find the same game in the betting datase.
bet_sample = bet_data[
    (bet_data["date"] == sample["date"])
    & ((bet_data["home"] == sample["club"]) | (bet_data["away"] == sample["club"]))
].iloc[0]

# Check if drawing odds is the same.
is_odds_equal = sample["draw"] == bet_sample["draw_odds"]
print("Merge was succesfull for this sample.\n")

# Visualize both samples
print("Cartola FC dataset sample:\n", sample, "\n")
print("Betting dataset sample:\n", bet_sample)

Merge was succesfull for this sample.

Cartola FC dataset sample:
 year                       2019
name            Filipe Trindade
round                        35
club                      Goiás
position               defender
status                     null
points                        0
mean                       0.75
date        2019-11-27 00:00:00
win                        8.19
lose                       1.37
draw                       4.67
Name: 58475, dtype: object 

Betting dataset sample:
 year                        2019
date         2019-11-27 00:00:00
home               Internacional
away                       Goiás
home_odds                   1.37
draw_odds                   4.67
away_odds                   8.19
club                       Goiás
Name: 3001, dtype: object


It confirms that pd.merge_asof was correctly set.

With that, we are done with the data wrangling process (finally). Now we can proceed to the fun part, the model creation.